File: ruby-service/app/circuit_breaker.rb
# frozen_string_literal: true

require 'json'
require 'net/http'
require 'uri'
require 'thread'
require 'monitor'

module CircuitBreaker
  class Error < StandardError; end
  class OpenError < Error
    attr_reader :name, :remaining_time

    def initialize(name, remaining_time)
      @name = name
      @remaining_time = remaining_time
      super("Circuit breaker '#{name}' is open. Retry after #{remaining_time.round(2)}s")
    end
  end

  class State
    CLOSED = :closed
    OPEN = :open
    HALF_OPEN = :half_open

    def self.all
      [CLOSED, OPEN, HALF_OPEN]
    end
  end

  class Config
    attr_accessor :failure_threshold, :success_threshold, :timeout_seconds,
                  :half_open_max_calls, :sliding_window_size, :failure_rate_threshold

    def initialize(
      failure_threshold: 5,
      success_threshold: 3,
      timeout_seconds: 30.0,
      half_open_max_calls: 3,
      sliding_window_size: 10,
      failure_rate_threshold: 0.5
    )
      @failure_threshold = failure_threshold
      @success_threshold = success_threshold
      @timeout_seconds = timeout_seconds
      @half_open_max_calls = half_open_max_calls
      @sliding_window_size = sliding_window_size
      @failure_rate_threshold = failure_rate_threshold
    end
  end

  class Metrics
    attr_reader :total_calls, :successful_calls, :failed_calls, :rejected_calls,
                :state_transitions, :last_failure_time, :last_success_time

    def initialize
      @total_calls = 0
      @successful_calls = 0
      @failed_calls = 0
      @rejected_calls = 0
      @state_transitions = 0
      @last_failure_time = nil
      @last_success_time = nil
      @response_times = []
      @max_response_times = 100
      @mutex = Mutex.new
    end

    def record_success(duration)
      @mutex.synchronize do
        @successful_calls += 1
        @total_calls += 1
        @last_success_time = Time.now
        add_response_time(duration)
      end
    end

    def record_failure(duration)
      @mutex.synchronize do
        @failed_calls += 1
        @total_calls += 1
        @last_failure_time = Time.now
        add_response_time(duration)
      end
    end

    def record_rejection
      @mutex.synchronize { @rejected_calls += 1 }
    end

    def record_state_transition
      @mutex.synchronize { @state_transitions += 1 }
    end

    def average_response_time
      @mutex.synchronize do
        return 0 if @response_times.empty?
        @response_times.sum / @response_times.size
      end
    end

    def to_h
      @mutex.synchronize do
        {
          total_calls: @total_calls,
          successful_calls: @successful_calls,
          failed_calls: @failed_calls,
          rejected_calls: @rejected_calls,
          state_transitions: @state_transitions,
          average_response_time_ms: (average_response_time * 1000).round(2),
          last_failure_time: @last_failure_time&.iso8601,
          last_success_time: @last_success_time&.iso8601
        }
      end
    end

    private

    def add_response_time(duration)
      @response_times << duration
      @response_times.shift if @response_times.size > @max_response_times
    end
  end

  class Breaker
    include MonitorMixin

    attr_reader :name, :config, :metrics

    @@registry = {}
    @@registry_mutex = Mutex.new

    def self.get_or_create(name, config: nil)
      @@registry_mutex.synchronize do
        @@registry[name] ||= new(name, config: config)
      end
    end

    def self.registry
      @@registry_mutex.synchronize { @@registry.dup }
    end

    def initialize(name, config: nil)
      super()
      @name = name
      @config = config || Config.new
      @state = State::CLOSED
      @failure_count = 0
      @success_count = 0
      @half_open_calls = 0
      @opened_at = nil
      @metrics = Metrics.new
      @sliding_window = Array.new(@config.sliding_window_size, true)
      @window_index = 0
    end

    def execute(fallback: nil, &block)
      raise ArgumentError, 'Block required' unless block_given?

      unless allow_request?
        @metrics.record_rejection
        return fallback.call if fallback
        remaining = @config.timeout_seconds - (Time.now - (@opened_at || Time.now))
        raise OpenError.new(@name, [remaining, 0].max)
      end

      start_time = Time.now
      begin
        result = yield
        duration = Time.now - start_time
        record_success(duration)
        result
      rescue StandardError => e
        duration = Time.now - start_time
        record_failure(duration)
        raise
      end
    end

    def state
      synchronize do
        if @state == State::OPEN && should_attempt_reset?
          transition_to(State::HALF_OPEN)
        end
        @state
      end
    end

    def health_info
      synchronize do
        {
          name: @name,
          state: @state.to_s.upcase,
          failure_count: @failure_count,
          success_count: @success_count,
          failure_rate: calculate_failure_rate,
          metrics: @metrics.to_h,
          config: {
            failure_threshold: @config.failure_threshold,
            success_threshold: @config.success_threshold,
            timeout_seconds: @config.timeout_seconds
          }
        }
      end
    end

    private

    def allow_request?
      synchronize do
        current_state = state

        case current_state
        when State::CLOSED
          true
        when State::OPEN
          false
        when State::HALF_OPEN
          if @half_open_calls < @config.half_open_max_calls
            @half_open_calls += 1
            true
          else
            false
          end
        else
          false
        end
      end
    end

    def should_attempt_reset?
      return false unless @opened_at
      Time.now - @opened_at >= @config.timeout_seconds
    end

    def transition_to(new_state)
      old_state = @state
      return if old_state == new_state

      @state = new_state
      @metrics.record_state_transition

      case new_state
      when State::OPEN
        @opened_at = Time.now
      when State::HALF_OPEN
        @half_open_calls = 0
        @success_count = 0
      when State::CLOSED
        @failure_count = 0
        @success_count = 0
        @opened_at = nil
        @sliding_window.fill(true)
        @window_index = 0
      end
    end

    def record_success(duration)
      synchronize do
        @metrics.record_success(duration)
        add_to_sliding_window(true)

        case @state
        when State::HALF_OPEN
          @success_count += 1
          transition_to(State::CLOSED) if @success_count >= @config.success_threshold
        when State::CLOSED
          @failure_count = [@failure_count - 1, 0].max
        end
      end
    end

    def record_failure(duration)
      synchronize do
        @metrics.record_failure(duration)
        add_to_sliding_window(false)

        case @state
        when State::HALF_OPEN
          transition_to(State::OPEN)
        when State::CLOSED
          @failure_count += 1
          failure_rate = calculate_failure_rate

          if @failure_count >= @config.failure_threshold ||
             failure_rate >= @config.failure_rate_threshold
            transition_to(State::OPEN)
          end
        end
      end
    end

    def add_to_sliding_window(success)
      @sliding_window[@window_index] = success
      @window_index = (@window_index + 1) % @config.sliding_window_size
    end

    def calculate_failure_rate
      failures = @sliding_window.count(false)
      failures.to_f / @sliding_window.size
    end
  end

  class DistributedCoordinator
    def initialize(coordinator_url, sync_interval: 5.0)
      @coordinator_url = coordinator_url
      @sync_interval = sync_interval
      @node_id = ENV['NODE_ID'] || "ruby-#{Process.pid}"
      @breakers = {}
      @running = false
      @sync_thread = nil
      @mutex = Mutex.new
    end

    def register(breaker)
      @mutex.synchronize { @breakers[breaker.name] = breaker }
      send_registration(breaker)
    end

    def start_sync
      @running = true
      @sync_thread = Thread.new do
        while @running
          sleep @sync_interval
          synchronize_states
        end
      end
    end

    def stop_sync
      @running = false
      @sync_thread&.join(2)
    end

    def get_cluster_state(service_name)
      uri = URI("#{@coordinator_url}/circuit-breakers/#{service_name}/aggregate")
      response = Net::HTTP.get_response(uri)
      JSON.parse(response.body)
    rescue StandardError => e
      { error: e.message }
    end

    private

    def send_registration(breaker)
      uri = URI("#{@coordinator_url}/circuit-breakers/register")
      http = Net::HTTP.new(uri.host, uri.port)
      http.open_timeout = 5
      http.read_timeout = 5

      request = Net::HTTP::Post.new(uri)
      request['Content-Type'] = 'application/json'
      request.body = {
        service: breaker.name,
        node_id: @node_id,
        failure_threshold: breaker.config.failure_threshold,
        success_threshold: breaker.config.success_threshold
      }.to_json

      http.request(request)
    rescue StandardError
    end

    def synchronize_states
      @mutex.synchronize do
        @breakers.each do |name, breaker|
          report_state(name, breaker)
        end
      end
    end

    def report_state(name, breaker)
      uri = URI("#{@coordinator_url}/circuit-breakers/state")
      http = Net::HTTP.new(uri.host, uri.port)
      http.open_timeout = 5
      http.read_timeout = 5

      request = Net::HTTP::Post.new(uri)
      request['Content-Type'] = 'application/json'
      request.body = {
        service: name,
        node_id: @node_id,
        state: breaker.state.to_s.upcase,
        timestamp: (Time.now.to_f * 1000).to_i,
        health_info: breaker.health_info
      }.to_json

      http.request(request)
    rescue StandardError
    end
  end
end

File: js-service/src/circuit-breaker.ts
export enum CircuitState {
  CLOSED = 'CLOSED',
  OPEN = 'OPEN',
  HALF_OPEN = 'HALF_OPEN',
}

export interface CircuitBreakerConfig {
  failureThreshold: number;
  successThreshold: number;
  timeoutMs: number;
  halfOpenMaxCalls: number;
  slidingWindowSize: number;
  failureRateThreshold: number;
}

export interface CircuitBreakerMetrics {
  totalCalls: number;
  successfulCalls: number;
  failedCalls: number;
  rejectedCalls: number;
  stateTransitions: number;
  lastFailureTime: Date | null;
  lastSuccessTime: Date | null;
  averageResponseTimeMs: number;
}

export interface HealthInfo {
  name: string;
  state: CircuitState;
  failureCount: number;
  successCount: number;
  failureRate: number;
  metrics: CircuitBreakerMetrics;
  config: Partial<CircuitBreakerConfig>;
}

const DEFAULT_CONFIG: CircuitBreakerConfig = {
  failureThreshold: 5,
  successThreshold: 3,
  timeoutMs: 30000,
  halfOpenMaxCalls: 3,
  slidingWindowSize: 10,
  failureRateThreshold: 0.5,
};

export class CircuitBreakerOpenError extends Error {
  constructor(
    public readonly name: string,
    public readonly remainingTimeMs: number
  ) {
    super(
      `Circuit breaker '${name}' is open. Retry after ${Math.round(remainingTimeMs)}ms`
    );
    this.name = 'CircuitBreakerOpenError';
  }
}

export class CircuitBreaker {
  private static registry = new Map<string, CircuitBreaker>();

  private state: CircuitState = CircuitState.CLOSED;
  private failureCount = 0;
  private successCount = 0;
  private halfOpenCalls = 0;
  private openedAt: number | null = null;
  private slidingWindow: boolean[];
  private windowIndex = 0;
  private responseTimes: number[] = [];
  private readonly maxResponseTimes = 100;

  private metrics: CircuitBreakerMetrics = {
    totalCalls: 0,
    successfulCalls: 0,
    failedCalls: 0,
    rejectedCalls: 0,
    stateTransitions: 0,
    lastFailureTime: null,
    lastSuccessTime: null,
    averageResponseTimeMs: 0,
  };

  private readonly config: CircuitBreakerConfig;

  constructor(
    public readonly name: string,
    config: Partial<CircuitBreakerConfig> = {}
  ) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.slidingWindow = new Array(this.config.slidingWindowSize).fill(true);
  }

  static getOrCreate(
    name: string,
    config?: Partial<CircuitBreakerConfig>
  ): CircuitBreaker {
    if (!this.registry.has(name)) {
      this.registry.set(name, new CircuitBreaker(name, config));
    }
    return this.registry.get(name)!;
  }

  static getRegistry(): Map<string, CircuitBreaker> {
    return new Map(this.registry);
  }

  async execute<T>(
    operation: () => Promise<T>,
    fallback?: () => Promise<T>
  ): Promise<T> {
    if (!this.allowRequest()) {
      this.metrics.rejectedCalls++;
      if (fallback) {
        return fallback();
      }
      const remaining = this.config.timeoutMs - (Date.now() - (this.openedAt ?? 0));
      throw new CircuitBreakerOpenError(this.name, Math.max(0, remaining));
    }

    const startTime = Date.now();
    this.metrics.totalCalls++;

    try {
      const result = await operation();
      const duration = Date.now() - startTime;
      this.recordSuccess(duration);
      return result;
    } catch (error) {
      const duration = Date.now() - startTime;
      this.recordFailure(duration);
      throw error;
    }
  }

  executeSync<T>(operation: () => T, fallback?: () => T): T {
    if (!this.allowRequest()) {
      this.metrics.rejectedCalls++;
      if (fallback) {
        return fallback();
      }
      const remaining = this.config.timeoutMs - (Date.now() - (this.openedAt ?? 0));
      throw new CircuitBreakerOpenError(this.name, Math.max(0, remaining));
    }

    const startTime = Date.now();
    this.metrics.totalCalls++;

    try {
      const result = operation();
      const duration = Date.now() - startTime;
      this.recordSuccess(duration);
      return result;
    } catch (error) {
      const duration = Date.now() - startTime;
      this.recordFailure(duration);
      throw error;
    }
  }

  getState(): CircuitState {
    if (this.state === CircuitState.OPEN && this.shouldAttemptReset()) {
      this.transitionTo(CircuitState.HALF_OPEN);
    }
    return this.state;
  }

  getHealthInfo(): HealthInfo {
    return {
      name: this.name,
      state: this.getState(),
      failureCount: this.failureCount,
      successCount: this.successCount,
      failureRate: this.calculateFailureRate(),
      metrics: { ...this.metrics },
      config: {
        failureThreshold: this.config.failureThreshold,
        successThreshold: this.config.successThreshold,
        timeoutMs: this.config.timeoutMs,
      },
    };
  }

  private allowRequest(): boolean {
    const currentState = this.getState();

    switch (currentState) {
      case CircuitState.CLOSED:
        return true;
      case CircuitState.OPEN:
        return false;
      case CircuitState.HALF_OPEN:
        if (this.halfOpenCalls < this.config.halfOpenMaxCalls) {
          this.halfOpenCalls++;
          return true;
        }
        return false;
      default:
        return false;
    }
  }

  private shouldAttemptReset(): boolean {
    if (this.openedAt === null) return false;
    return Date.now() - this.openedAt >= this.config.timeoutMs;
  }

  private transitionTo(newState: CircuitState): void {
    if (this.state === newState) return;

    this.state = newState;
    this.metrics.stateTransitions++;

    switch (newState) {
      case CircuitState.OPEN:
        this.openedAt = Date.now();
        break;
      case CircuitState.HALF_OPEN:
        this.halfOpenCalls = 0;
        this.successCount = 0;
        break;
      case CircuitState.CLOSED:
        this.failureCount = 0;
        this.successCount = 0;
        this.openedAt = null;
        this.slidingWindow.fill(true);
        this.windowIndex = 0;
        break;
    }
  }

  private recordSuccess(duration: number): void {
    this.metrics.successfulCalls++;
    this.metrics.lastSuccessTime = new Date();
    this.addResponseTime(duration);
    this.addToSlidingWindow(true);

    switch (this.state) {
      case CircuitState.HALF_OPEN:
        this.successCount++;
        if (this.successCount >= this.config.successThreshold) {
          this.transitionTo(CircuitState.CLOSED);
        }
        break;
      case CircuitState.CLOSED:
        this.failureCount = Math.max(0, this.failureCount - 1);
        break;
    }
  }

  private recordFailure(duration: number): void {
    this.metrics.failedCalls++;
    this.metrics.lastFailureTime = new Date();
    this.addResponseTime(duration);
    this.addToSlidingWindow(false);

    switch (this.state) {
      case CircuitState.HALF_OPEN:
        this.transitionTo(CircuitState.OPEN);
        break;
      case CircuitState.CLOSED:
        this.failureCount++;
        const failureRate = this.calculateFailureRate();

        if (
          this.failureCount >= this.config.failureThreshold ||
          failureRate >= this.config.failureRateThreshold
        ) {
          this.transitionTo(CircuitState.OPEN);
        }
        break;
    }
  }

  private addToSlidingWindow(success: boolean): void {
    this.slidingWindow[this.windowIndex] = success;
    this.windowIndex = (this.windowIndex + 1) % this.config.slidingWindowSize;
  }

  private calculateFailureRate(): number {
    const failures = this.slidingWindow.filter((s) => !s).length;
    return failures / this.slidingWindow.length;
  }

  private addResponseTime(duration: number): void {
    this.responseTimes.push(duration);
    if (this.responseTimes.length > this.maxResponseTimes) {
      this.responseTimes.shift();
    }
    this.metrics.averageResponseTimeMs =
      this.responseTimes.reduce((a, b) => a + b, 0) / this.responseTimes.length;
  }
}

export interface AggregatedState {
  service: string;
  consensusState: CircuitState;
  totalNodes: number;
  healthScore: number;
  nodeStates: Record<string, CircuitState>;
}

export class DistributedCircuitBreakerClient {
  private breakers = new Map<string, CircuitBreaker>();
  private syncInterval: ReturnType<typeof setInterval> | null = null;
  private readonly nodeId: string;

  constructor(
    private readonly coordinatorUrl: string,
    private readonly syncIntervalMs: number = 5000
  ) {
    this.nodeId = process.env.NODE_ID ?? `ts-${process.pid}`;
  }

  register(breaker: CircuitBreaker): void {
    this.breakers.set(breaker.name, breaker);
    this.sendRegistration(breaker).catch(() => { });
  }

  startSync(): void {
    if (this.syncInterval) return;

    this.syncInterval = setInterval(() => {
      this.synchronizeStates().catch(() => { });
    }, this.syncIntervalMs);
  }

  stopSync(): void {
    if (this.syncInterval) {
      clearInterval(this.syncInterval);
      this.syncInterval = null;
    }
  }

  async getAggregatedState(serviceName: string): Promise<AggregatedState> {
    try {
      const response = await fetch(
        `${this.coordinatorUrl}/circuit-breakers/${serviceName}/aggregate`
      );
      return (await response.json()) as AggregatedState;
    } catch {
      return {
        service: serviceName,
        consensusState: CircuitState.CLOSED,
        totalNodes: 0,
        healthScore: 0,
        nodeStates: {},
      };
    }
  }

  private async sendRegistration(breaker: CircuitBreaker): Promise<void> {
    await fetch(`${this.coordinatorUrl}/circuit-breakers/register`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        service: breaker.name,
        node_id: this.nodeId,
        failure_threshold: breaker.getHealthInfo().config.failureThreshold,
        success_threshold: breaker.getHealthInfo().config.successThreshold,
      }),
    });
  }

  private async synchronizeStates(): Promise<void> {
    for (const [name, breaker] of this.breakers) {
      await this.reportState(name, breaker).catch(() => { });
    }
  }

  private async reportState(name: string, breaker: CircuitBreaker): Promise<void> {
    await fetch(`${this.coordinatorUrl}/circuit-breakers/state`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        service: name,
        node_id: this.nodeId,
        state: breaker.getState(),
        timestamp: Date.now(),
        health_info: breaker.getHealthInfo(),
      }),
    });
  }
}

export function withCircuitBreaker<T extends (...args: any[]) => Promise<any>>(
  name: string,
  config?: Partial<CircuitBreakerConfig>
) {
  const breaker = CircuitBreaker.getOrCreate(name, config);

  return function decorator(
    target: any,
    propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;

    descriptor.value = async function (...args: any[]) {
      return breaker.execute(() => originalMethod.apply(this, args));
    };

    return descriptor;
  };
}
